store/s3.go

```diff
@@ -131,13 +139,44 @@ func (s *S3) Delete(key string) error {
 		Bucket: aws.String(s.Bucket),
 		Key:    aws.String(s.Prefix + key),
 	})
+	if err != nil {
+		log.Println("S3 Delete:", s.Prefix, key, err)
+		raven.CaptureError(err, map[string]string{"Bucket": s.Bucket, "Prefix": s.Prefix, "Key": key})
+	} else {
+		s.setkeysize(key, -1)
+	}
 	return err
 }
 
 // stat will check if a key exists, and if so it returns the size. If the item
 // does not exist an error is returned. The prefix is added to the key before
 // checking.
 func (s *S3) stat(key string) (int64, error) {
+	// Cache the key sizes as we see them. This drastically cuts down on the
+	// number of HEAD requests.
+	s.m.Lock()
+	defer s.m.Unlock()
+	entry := s.cache[key]
+	if entry.size > 0 {
+		return entry.size, nil
+	}
+	if entry.size < 0 {
+		// we have previously determined this key does not exist
+		return 0, ErrNotExist
+	}
+	// key is not cached, so do the HEAD request
+	size, err := s.stat0(key)
+	if err != nil {
+		s.cache[key] = head{ttl: defaultMissTTL, size: size}
```

Comment:
Why is this not using `setkeysize` from below?

Reply:
i know! the problem is the `m` lock. the code here is holding it. But `setkeysize` also tries to grab it. hmm. ok, yeah, they should be unified.